### PR TITLE
Switch to Github Actions

### PR DESCRIPTION
**Pull request by afranke**
This is fairly similar to how it is done with Buildkite.

I used a more modern and pythonic approach in places (e.g. f-strings, and accessing dictionary items with brackets instead of `get`) for readability. It leads to some style inconsistencies with the existing code, but I didn’t want to touch the bits that don’t need to be changed for the purpose of this PR. I can do another pass later to make it more consistent if that is desired.

### CODE REVIEW

File: gha-webhook-listener.py [hunk 1]
```diff
@@ -64,89 +70,88 @@ def create_symlink(source, linkname):
...
+    logger.info("Fetching %s", artifacts_url)
     artifacts_resp = requests.get(artifacts_url, headers=req_headers())
     artifacts_resp.raise_for_status()
-    artifacts_array = artifacts_resp.json()
+    artifacts_array = artifacts_resp.json()['artifacts']
 
     artifact_to_deploy = None
     for artifact in artifacts_array:
-        if re.match(arg_artifact_pattern, artifact['path']):
+        logging.info("Artifact name is %s", artifact['name'])
```
Comment (richvdh) [inline gha-webhook-listener.py:145]: This is going to the wrong logger, and I think we can downgrade it to `debug` and improve the wording:

```suggestion
                logger.debug("Considering artifact %s", artifact['name'])
```

File: gha-webhook-listener.py [hunk 2]
```diff
@@ -159,11 +164,11 @@ def on_receive_buildkite_poke():
     #       a good deploy with a bad one
     #   (b) we'll be overwriting the live deployment, which means people might
     #       see half-written files.
-    target_dir = os.path.join(arg_extract_path, "%s-#%i" % (pipeline_name, build_num))
+    target_dir = os.path.join(arg_extract_path, "%s-#%i" % (workflow_id, build_id))
     if os.path.exists(target_dir):
         abort(400, "Not deploying. We have previously deployed this build.")
 
-    # buildkite times out the request if it takes longer than 10s, and fetching
+    # Github might time out the request if it takes longer than 10s, and fetching
```
Comment (richvdh) [inline gha-webhook-listener.py:171]:
```suggestion
    # Github might time out the request if it takes a long time, and fetching
```

File: gha-webhook-listener.py [hunk 3]
```diff
@@ -22,14 +23,19 @@
 from __future__ import print_function
 
 import argparse
+from dateutil import parser as dateparser
```
Comment (richvdh) [inline gha-webhook-listener.py:26]: this doesn't actually seem to be used? could you remove it, and remove the dependency from `setup.py`

File: gha-webhook-listener.py [hunk 4]
```diff
@@ -195,9 +200,14 @@ def deploy_tarball(artifact_url, target_dir):
     resp = requests.get(artifact_url, stream=True, headers=req_headers())
     resp.raise_for_status()
 
+    # GHA artifacts are wrapped in a zip file, so we extract it to get our tarball
+    # See https://github.com/actions/upload-artifact/issues/109
+    zipped_artifact = zipfile.ZipFile(BytesIO(resp.content))
+    tarball = zipped_artifact.open('content.tar.gz')
+
     # TODO: make the compression type depend on the filename
     # (or copy it to a temporary file so that tarfile can autodetect)
```
Comment (richvdh) [inline gha-webhook-listener.py:209]: we know what the filename is, because its hardcoded on the line above, so I think this TODO is redundant.

**Review by richvdh — CHANGES_REQUESTED** [review]
thanks for this! looks good, but a few things to clean up.

File: gha-webhook-listener.py [hunk 5]
```diff
@@ -2,10 +2,11 @@
 #
 # auto-deploy listener script
 #
-# Listens for buildkite webhook pokes. When it gets one, downloads the artifact
-# from buildkite and unpacks it.
+# Listens for Github webhook pokes. When it gets one, downloads the artifact
+# from Github and unpacks it.
 #
 # Copyright 2019 New Vector Ltd
+# Copyright 2021 The Matrix.org Foundation C.I.C.
```
[the remark this answers is not in the document]
Comment (afranke) [reply gha-webhook-listener.py:9]: Is this even legal, considering these are two different entities?

File: gha-webhook-listener.py (hunk 3 above)
Comment (afranke) [reply gha-webhook-listener.py:26]: Indeed, leftover from a previous iteration

File: gha-webhook-listener.py [hunk 6]
```diff
@@ -249,27 +259,27 @@ def tidy_extract_directory(target_dir, cleanup_dir, versions_to_keep):
 
     parser.add_argument(
         "--webhook-token", dest="webhook_token", help=(
-            "Only accept pokes with this buildkite token."
+            "Only accept pokes with this Github token."
         ), required=True,
     )
```
[the remark this answers is not in the document]
Comment (afranke) [reply gha-webhook-listener.py]: A secret is an item that you can store alongside your repository and could take many forms (e.g. an SSH key). Here, we use a specific kind of secret which is a token, so the term is correct and better as it is more specific. See https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks#setting-your-secret-token

File: gha-webhook-listener.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply gha-webhook-listener.py:9]: good question. Apparently there were some legal contortions to transfer copyright for things in the matrix-org org from NVL to M.org:

image

[the remark this answers is not in the document]
Comment (afranke) [reply gha-webhook-listener.py:9]: Fair enough.

File: gha-webhook-listener.py [hunk 7]
```diff
@@ -64,89 +70,88 @@ def create_symlink(source, linkname):
 
 def req_headers():
     return {
-        "Authorization": "Bearer %s" % (arg_api_token,),
+        "Authorization": "token %s" % (arg_api_token,),
     }
 
 
+def validate_signature(data, signature, token):
```
[the remark this answers is not in the document]
Comment (afranke) [reply gha-webhook-listener.py]: Do the latest changes here work for you?

File: gha-webhook-listener.py [hunk 8]
```diff
@@ -74,21 +72,40 @@ def req_headers():
...
+    """
+    github_secret = token.encode('utf-8')
     expected_signature = hmac.new(key=github_secret, msg=data, digestmod=hashlib.sha256).hexdigest()
-    signature = signature.split('sha256=')[-1].strip()
-    return hmac.compare_digest(signature, expected_signature)
+    if not signature.startswith('sha256='):
+        return False
+    else:
+        signature = signature.split('sha256=')[-1].strip()
+        return hmac.compare_digest(signature, expected_signature)
```
Comment (richvdh) [inline gha-webhook-listener.py:93]:
```suggestion
    signature = signature[7:].strip()
    return hmac.compare_digest(signature, expected_signature)
```

File: gha-webhook-listener.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply gha-webhook-listener.py]: they certainly do!

File: setup.py [hunk 9]
```diff
@@ -18,13 +18,14 @@
 from setuptools import setup
 
 setup(
-    name="buildkite-webhook-listener",
+    name="gha-webhook-listener",
     install_requires=[
+        "python-dateutil",
```
Comment (richvdh) [inline setup.py:23]: please remove this if we're not using it.

**Review by richvdh — APPROVED** [review]
lgtm otherwise!

File: gha-webhook-listener.py (hunk 8 above)
Comment (afranke) [reply gha-webhook-listener.py:93]: Funny, I initially wrote like this. I don’t know why I changed it. :smile:

**Pull request merged**